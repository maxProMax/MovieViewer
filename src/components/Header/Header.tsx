import { FC, useEffect, useState } from 'react';
import { useLocation } from 'react-router-dom';
import MenuIcon from '@mui/icons-material/Menu';
import Drawer from '@mui/material/Drawer';
import { LinkHome } from 'src/router/links';
import { Navigation } from './Navigation';
import styles from './header.module.css';

export const Header: FC = () => {
    const [open, change] = useState(false);
    const location = useLocation();

    useEffect(() => {
        change(false);
    }, [location]);

    return (
        <header className={styles.container}>
            <div className={styles.wrapper}>
                <button className={styles.burger} onClick={() => change(true)}>
                    <MenuIcon />
                </button>
                <Drawer open={open} onClose={() => change(false)}>
                    <Navigation />
                </Drawer>
                <div className={styles.logo}>
                    <LinkHome>MovieViewer</LinkHome>
                </div>
            </div>
        </header>
    );
};
