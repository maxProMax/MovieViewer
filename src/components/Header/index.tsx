import { FC, useEffect, useState } from 'react';
import { useLocation } from 'react-router-dom';
import MenuIcon from '@mui/icons-material/Menu';
import Drawer from '@mui/material/Drawer';
import { Navigation } from './Navigation';
import { LinkHome } from '../../router/links';
import styles from './header.module.css';

export const Header: FC = () => {
    const [open, change] = useState(false);
    const location = useLocation();

    useEffect(() => {
        change(false);
    }, [location]);

    return (
        <header className={styles.container}>
            <div className={styles.wrapper}>
                <span className={styles.burger} onClick={() => change(true)}>
                    <MenuIcon />
                </span>
                <Drawer open={open} onClose={() => change(false)}>
                    <Navigation />
                </Drawer>
                <div className={styles.logo}>
                    <LinkHome>MovieViewer</LinkHome>
                </div>
            </div>
        </header>
    );
};
