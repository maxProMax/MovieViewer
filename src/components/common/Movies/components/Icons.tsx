import { FC, PropsWithChildren, useCallback } from 'react';
import FavoriteBorderIcon from '@mui/icons-material/FavoriteBorder';
import MuiFavoriteIcon from '@mui/icons-material/Favorite';
import BookmarkAddIcon from '@mui/icons-material/BookmarkAdd';
import BookmarkAddedIcon from '@mui/icons-material/BookmarkAdded';
import { useFavorites, useWatchLatter } from 'src/store/hooks';
import { useMovie } from 'src/hooks/api';
import styles from './icons.module.css';

const Button: FC<
    PropsWithChildren<{ disabled: boolean; onClick: () => void }>
> = (props) => (
    <button
        tabIndex={0}
        className={`${styles.icon} ${props.disabled ? styles.disabled : ''}`}
        {...props}
    />
);

export const FavoriteIcon: FC<{ movieId: number }> = ({ movieId }) => {
    const { add, remove, favorites } = useFavorites();
    const { request, loading } = useMovie();
    const added = Boolean(favorites[movieId]);

    const handleClick = useCallback(() => {
        added ? remove(movieId) : request(movieId).then((d) => d && add(d));
    }, [added]);

    return (
        <Button disabled={loading} onClick={handleClick}>
            {added ? <MuiFavoriteIcon /> : <FavoriteBorderIcon />}
        </Button>
    );
};

export const WatchLater: FC<{ movieId: number }> = ({ movieId }) => {
    const { add, remove, watchLater } = useWatchLatter();
    const { request, loading } = useMovie();
    const added = watchLater[movieId];

    const handleClick = useCallback(() => {
        added ? remove(movieId) : request(movieId).then((d) => d && add(d));
    }, [added]);

    return (
        <Button disabled={loading} onClick={handleClick}>
            {added ? <BookmarkAddedIcon /> : <BookmarkAddIcon />}
        </Button>
    );
};
