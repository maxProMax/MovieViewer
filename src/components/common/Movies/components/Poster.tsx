import { FC } from 'react';
import { ImageThemoviedb, Image } from '../../../common/Image';
import stylesMovie from './movie.module.css';

export const Poster: FC<{ path: string | null; title: string }> = ({
    path,
    title,
}) =>
    !path ? (
        <Image
            className={stylesMovie.image}
            src={'./image-not-found.png'}
            alt={title}
        />
    ) : (
        <ImageThemoviedb
            className={stylesMovie.image}
            path={path}
            alt={title}
        />
    );
