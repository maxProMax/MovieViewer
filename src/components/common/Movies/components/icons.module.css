.icon {
    display: inline-flex;
    cursor: pointer;
    background: none;
    border: none;
}

.disabled {
    cursor: initial;
    color: grey;
}
