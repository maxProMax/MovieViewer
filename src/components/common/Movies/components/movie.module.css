.movie {
    --border-radius: 4px;

    position: relative;
    display: flex;
    flex-direction: column;
    row-gap: 12px;
    border: 1px solid rgba(0, 0, 0, 0.23);
    padding: 8px;
    border-radius: var(--border-radius);
}

.title {
    display: flex;
    column-gap: 8px;
    align-items: center;
    color: #666;
}

.titleText {
    flex-grow: 1;
}

.image {
    width: 100%;
    border-radius: var(--border-radius);
}

.icons {
    display: flex;
    column-gap: 4px;
}
