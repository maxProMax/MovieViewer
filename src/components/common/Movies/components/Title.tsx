import { FC } from 'react';
import Typography from '@mui/material/Typography';
import { FavoriteIcon, WatchLater } from './Icons';
import stylesMovie from './movie.module.css';

export const Title: FC<{ title?: string; id?: number; className?: string }> = ({
    title,
    id,
    className = '',
}) =>
    !title ? null : (
        <Typography
            className={`${stylesMovie.title} ${className}`}
            variant="body1"
        >
            <strong className={stylesMovie.titleText}>{title}</strong>
            {id !== undefined && (
                <span className={stylesMovie.icons}>
                    <FavoriteIcon movieId={id} />
                    <WatchLater movieId={id} />
                </span>
            )}
        </Typography>
    );
