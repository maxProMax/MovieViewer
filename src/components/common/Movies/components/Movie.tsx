import { FC } from 'react';
import Typography from '@mui/material/Typography';
import { MovieShort as MovieType } from 'src/modules/apis/themoviedb/types';
import { LinkMovie } from 'src/router/links';
import { Poster } from 'src/components/common/Poster';
import { Title } from './Title';
import stylesMovie from './movie.module.css';

export const Movie: FC<{ movie: MovieType }> = ({ movie }) => {
    const { id, poster_path, title, overview } = movie;

    return (
        <div className={stylesMovie.movie} tabIndex={0}>
            <LinkMovie id={String(movie.id)}>
                <Poster path={poster_path} title={title} />
            </LinkMovie>
            <Title title={title} id={id} />
            <Typography variant="body2">
                {overview.length >= 170
                    ? `${overview.substring(0, 140)}...`
                    : overview}
            </Typography>
        </div>
    );
};
