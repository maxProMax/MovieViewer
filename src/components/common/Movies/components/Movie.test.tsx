import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { GlobalProvider } from 'src/Providers';
import { mockedFetch } from 'src/tests/mockedFetch';
import { Header } from 'src/components/Header';
import { MovieShort } from 'src/modules/apis/themoviedb/types';
import { Movie } from './Movie';
import { mockMovie } from 'src/tests/data';

const movie: MovieShort = {
    ...mockMovie,
    title: 'Movie_Title',
};

describe('Movie component', () => {
    beforeEach(() => {
        mockedFetch(() => Promise.resolve(movie));
    });

    test('Render', async () => {
        await render(<Movie movie={movie} />, { wrapper: GlobalProvider });

        expect(screen.getByRole('link')).toBeInTheDocument();
    });

    test.each([
        {
            name: 'Favorites',
            iconName: 'FavoriteBorderIcon',
            expected: /Favorites\s\(\d\)/,
        },
        {
            name: 'Watch Later',
            iconName: 'BookmarkAddIcon',
            expected: /Watch\sLater\s\(\d\)/,
        },
    ])('Add to $name', async ({ iconName, expected }) => {
        await render(
            <GlobalProvider>
                <Header />
                <Movie movie={movie} />
            </GlobalProvider>
        );

        const favoriteIcon = screen.getByTestId(iconName);

        expect(favoriteIcon).toBeInTheDocument();

        userEvent.click(favoriteIcon);

        const menuIcon = screen.getByTestId('MenuIcon');

        await userEvent.click(menuIcon);

        const link = await screen.findByRole('link', { name: expected });

        expect(link).toBeInTheDocument();
    });

    test('layout', async () => {
        const view = render(<Movie movie={movie} />, {
            wrapper: GlobalProvider,
        });

        expect(view.container).toMatchSnapshot();
    });
});
