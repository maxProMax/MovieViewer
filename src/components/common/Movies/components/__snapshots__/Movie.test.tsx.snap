// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie component layout 1`] = `
<div>
  <div
    class="movie"
    tabindex="0"
  >
    <a
      class="link"
      href="/movie/1"
      id="1"
    >
      <picture>
        <source
          media="(min-width:780px)"
          srcset="undefinedw780/4I6IJwKxjY1HTEkfLmlp4aWoflT.jpg"
        />
        <source
          media="(min-width:342px)"
          srcset="undefinedw342/4I6IJwKxjY1HTEkfLmlp4aWoflT.jpg"
        />
        <img
          alt="Movie_Title"
          class="image "
          src="undefinedoriginal/4I6IJwKxjY1HTEkfLmlp4aWoflT.jpg"
        />
      </picture>
    </a>
    <p
      class="MuiTypography-root MuiTypography-body1 title  css-ahj2mt-MuiTypography-root"
    >
      <strong
        class="titleText"
      >
        Movie_Title
      </strong>
      <span
        class="icons"
      >
        <button
          class="icon "
          tabindex="0"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="FavoriteIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="m12 21.35-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
            />
          </svg>
        </button>
        <button
          class="icon "
          tabindex="0"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="BookmarkAddedIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="m19 21-7-3-7 3V5c0-1.1.9-2 2-2h7c-.63.84-1 1.87-1 3 0 2.76 2.24 5 5 5 .34 0 .68-.03 1-.1V21zM17.83 9 15 6.17l1.41-1.41 1.41 1.41 3.54-3.54 1.41 1.41L17.83 9z"
            />
          </svg>
        </button>
      </span>
    </p>
    <p
      class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
    >
      Take a trip back to the 90's in Kolkata
    </p>
  </div>
</div>
`;
